keps/sig-cluster-lifecycle/kubeadm/2915-kubeadm-replace-kubelet-config-x.y/README.md

```diff
@@ -0,0 +1,606 @@
+
+- Apply the change within a single release of kubeadm, breaking the users.
+
+## Proposal
+
+<!--
+This is where we get down to the specifics of what the proposal actually is.
+This should have enough detail that reviewers can understand exactly what
+you're proposing, but should not include things like API designs or
+implementation. What is the desired outcome and how do we measure success?.
+The "Design Details" section below is for the real
+nitty-gritty.
+-->
+
+The proposal is to declare the default ConfigMap and RBAC object format of `kubelet-config-x.y`
+deprecated but support it for a GA deprecation period of 1 year (3 kubeadm releases). In parallel,
+introduce objects named `kubelet-config` (without `x.y`) that will become the source of truth.
+
+The switch will be controlled by a new kubeadm specific feature gate (not a core one) called
+`UnversionedKubeletConfigMap`. Using a feature gate for similar changes is common in changes
+in core Kubernetes components and is familiar mechanism to users.
+
+Note that depending on feedback from users the N+x release schedule might be adjusted - e.g.
+extending or shortening the Alpha or Beta periods.
+
+#### First release N (when the deprecation happens)
+
+- The feature gate is added as Alpha and disabled by default.
+- Users can continue using the old naming format or opt-in into the new naming format
+by enabling the feature gate.
+
+#### Release N+1
+
+- The feature gate becomes Beta and is enabled by default.
+- Users can opt-out of the new feature by disabling the feature gate.
+
+#### Release N+3 (two releases after Beta)
+
+- The feature gate becomes GA and is locked to enabled.
+- Users can no longer opt-out and must adapt their infrastructure that reads/writes from/to the ConfigMap.
+- Users should remove usage of the feature gate from their infrastructure. Alternatively they can do it
+in the next release.
+
+#### Release N+4
+
+- The feature gate is removed.
+
+### User Stories (Optional)
+
+<!--
+Detail the things that people will be able to do if this KEP is implemented.
+Include as much detail as possible so that people can understand the "how" of
+the system. The goal here is to make this feel real for users without getting
+bogged down.
+-->
+
+#### Story 1
+
+As a kubeadm user that wishes to modify the KubeletConfiguration that kubeadm stores in the cluster
+before joining new nodes, I wish to have easy access to a static location such as `kube-system/kubelet-config`
+without looking up the control plane version first.
+
+#### Story 2
+
+As a kubeadm maintainer, I wish to simplify the logic in kubeadm that manages the kubelet configuration
+read/write operations.
+
+### Notes/Constraints/Caveats (Optional)
+
+<!--
+What are the caveats to the proposal?
+What are some important details that didn't come across above?
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+NONE
+
+### Risks and Mitigations
```

Comment:
I wold like a third risk, being kubelet starting to use a new API version for its own configuration, but not providing automatic conversion in place. This will prevent kubeadm to support kubelet N and kubelet N-1 at the same time.

A viable path forward is to use different keys inside the same ConfigMap; in order to make this possible in the future I suggest to start using API version as key in the new ConfigMap instead of kubelet.

Reply:
i can add this risk in the list.

i think an easier way for us to mitigate would be to ensure that we do not store the latest kubelet API version in the config map for one more release.

for example:
- kubelet API v1beta2 comes out in 1.26.
- for new clusters, kubeadm 1.26 should not write v1beta2 in the CM and should write v1beta1 instead.
- this would allow 1.25 kubelets to join the cluster as well as both 1.26 and 1.25 kubelets would still support v1beta1.

once 1.27 is out (the v1beta2 has been supported for 2 releases):
- new clusters created with kubeadm 1.27 can start writing v1beta2 in the CM.
- 1.26 kubelets would support v1beta2 and they can join the cluster.

i expect sig-node to listen to the requests to support converters on the kubelet CLI (or endpoint) once the v1beta1 is marked as deprecated.

Reply:
added as "Risk 3", as per my alternative approach above.